sklearn_numba_dpex/kmeans/kernels.py

```diff
@@ -0,0 +1,745 @@
+import math
+from functools import lru_cache
+
+import numba_dpex as dpex
+import dpctl
+
+
+@lru_cache
+def make_initialize_to_zeros_1dim_kernel(n_samples, work_group_size, dtype):
+
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        x[sample_idx] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_1dim_kernel(n_samples, work_group_size, dtype):
+
+    global_size = math.ceil(n_samples / work_group_size) * work_group_size
+    zero = dtype(0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        sample_idx = dpex.get_global_id(0)
+
+        if sample_idx >= n_samples:
+            return
+
+        x[sample_idx] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+
+    nb_items = n_samples * n_features
+    global_size = math.ceil(nb_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= nb_items:
+            return
+
+        i = item_idx // n_samples
+        j = item_idx % n_samples
+        x[i, j] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_2dim_kernel(n_samples, n_features, work_group_size, dtype):
+
+    nb_items = n_samples * n_features
+    global_size = math.ceil(nb_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= nb_items:
+            return
+
+        i = item_idx // n_samples
+        j = item_idx % n_samples
+        x[i, j] = zero
+
+    return initialize_to_zeros[global_size, work_group_size]
+
+
+@lru_cache
+def make_initialize_to_zeros_3dim_kernel(dim0, dim1, dim2, work_group_size, dtype):
+
+    nb_items = dim0 * dim1 * dim2
+    stride0 = dim1 * dim2
+    global_size = math.ceil(nb_items / work_group_size) * work_group_size
+    zero = dtype(0.0)
+
+    @dpex.kernel
+    def initialize_to_zeros(x):
+        item_idx = dpex.get_global_id(0)
+
+        if item_idx >= nb_items:
+            return
+
+        i = item_idx // stride0
+        stride0_id = item_idx % stride0
+        j = stride0_id // dim2
+        k = stride0_id % dim2
+        x[i, j, k] = zero
```

Comment:
No I didn't benchmark several alternatives for secondary kernels.

I did a quick timeit of each step of the kmeans to ensure that there weren't any major misbehavior and fixed if I found unexpected execution times.

In the end I don't think those "support kernels" put together hold for more than 20% of the total execution time of the KMeans (and this might already a generous take).

I too am really interested in what we could learn from gpu programming with those benchmarks, but regarding the goal of improving the KMeans it will be marginal so I thought that could be left as secondary TODOs (if ever).